File: engine/src/tunnel.ts
/**
 * Tube Tunnel — Spawns external tunnel processes (ngrok, tailscale funnel, cloudflared).
 *
 * Each tunnel type spawns a child process, captures the public URL from stdout,
 * and exposes it through the WebSocket API.
 */

import { spawn, type ChildProcess } from "node:child_process";

// ─── Types ───────────────────────────────────────────────────────────────────

export type TunnelType = "ngrok" | "tailscale" | "funnel";

export interface TunnelStatus {
  ngrok: boolean;
  tailscale: boolean;
  funnel: boolean;
  urls: Record<string, string>;
}

interface ActiveTunnel {
  process: ChildProcess;
  url?: string;
}

// ─── State ───────────────────────────────────────────────────────────────────

const tunnels = new Map<TunnelType, ActiveTunnel>();

// ─── URL extraction (per-tunnel-type stdout parser) ─────────────────────────

const URL_PATTERNS: Record<TunnelType, RegExp> = {
  ngrok: /https:\/\/[a-zA-Z0-9-]+\.ngrok-free\.app/,
  tailscale: /https:\/\/[a-z0-9-]+(?:-[a-z0-9]+)*\.(?:ts\.net|tailscale\.net)/i,
  funnel: /https:\/\/[a-zA-Z0-9-]+\.trycloudflare\.com/,
};

function extractUrl(type: TunnelType, text: string): string | undefined {
  const m = text.match(URL_PATTERNS[type]);
  return m ? m[0] : undefined;
}

// ─── Spawn / Kill ────────────────────────────────────────────────────────────

export function startTunnel(
  type: TunnelType,
  proxyPort: number,
  onUpdate: (status: TunnelStatus) => void
): void {
  if (tunnels.has(type)) return;

  const args = ((): [string, string[]] => {
    switch (type) {
      case "ngrok":
        return ["ngrok", ["http", String(proxyPort), "--log=stdout"]];
      case "tailscale":
        return ["tailscale", ["funnel", "--bg", String(proxyPort)]];
      case "funnel":
        return ["cloudflared", ["tunnel", "--no-autoupdate", "--url", `http://localhost:${proxyPort}`]];
    }
  })();

  const proc = spawn(args[0], args[1], {
    stdio: ["ignore", "pipe", "pipe"],
  });

  const entry: ActiveTunnel = { process: proc };
  tunnels.set(type, entry);

  let buffer = "";

  const onData = (chunk: Buffer) => {
    buffer += chunk.toString();
    const url = extractUrl(type, buffer);
    if (url && url !== entry.url) {
      entry.url = url;
      console.error(`[tube] Tunnel ${type}: ${url}`);
      onUpdate(getStatus());
    }
  };

  proc.stdout?.on("data", onData);
  proc.stderr?.on("data", onData);

  proc.on("exit", () => {
    tunnels.delete(type);
    onUpdate(getStatus());
  });

  proc.on("error", (err) => {
    console.error(`[tube] Tunnel ${type} spawn error: ${err.message}`);
    tunnels.delete(type);
    onUpdate(getStatus());
  });

  console.error(`[tube] Tunnel ${type}: started`);
}

export function stopTunnel(type: TunnelType): void {
  const entry = tunnels.get(type);
  if (!entry) return;
  entry.process.kill("SIGTERM");
  // Also try SIGKILL after 3s if still alive
  setTimeout(() => {
    try { entry.process.kill("SIGKILL"); } catch {}
  }, 3000);
  tunnels.delete(type);
  console.error(`[tube] Tunnel ${type}: stopped`);
}

export function getStatus(): TunnelStatus {
  const urls: Record<string, string> = {};
  for (const [type, entry] of tunnels) {
    if (entry.url) urls[type] = entry.url;
  }
  return {
    ngrok: tunnels.has("ngrok"),
    tailscale: tunnels.has("tailscale"),
    funnel: tunnels.has("funnel"),
    urls,
  };
}

File: proxy/tunnel.go
package proxy

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"sync"
)

// TunnelType identifies the tunnel provider.
type TunnelType string

const (
	TunnelNgrok     TunnelType = "ngrok"
	TunnelTailscale TunnelType = "tailscale"
	TunnelFunnel    TunnelType = "funnel"
)

// TunnelStatus represents active tunnel state.
type TunnelStatus struct {
	Ngrok     bool              `json:"ngrok"`
	Tailscale bool              `json:"tailscale"`
	Funnel    bool              `json:"funnel"`
	URLs      map[string]string `json:"urls"`
}

type activeTunnel struct {
	cmd *exec.Cmd
	url string
}

// TunnelManager spawns and manages tunnel processes.
type TunnelManager struct {
	mu       sync.Mutex
	tunnels  map[TunnelType]*activeTunnel
	onChange func(TunnelStatus)
}

var urlPatterns = map[TunnelType]*regexp.Regexp{
	TunnelNgrok:     regexp.MustCompile(`https://[a-zA-Z0-9-]+\.ngrok-free\.app`),
	TunnelTailscale: regexp.MustCompile(`https://[a-z0-9][a-z0-9.-]*\.(?:ts\.net|tailscale\.net)`),
	TunnelFunnel:    regexp.MustCompile(`https://[a-zA-Z0-9-]+\.trycloudflare\.com`),
}

// NewTunnelManager creates a tunnel manager.
func NewTunnelManager() *TunnelManager {
	return &TunnelManager{
		tunnels: make(map[TunnelType]*activeTunnel),
	}
}

// SetOnChange sets a callback that fires when tunnel state changes.
func (m *TunnelManager) SetOnChange(fn func(TunnelStatus)) {
	m.onChange = fn
}

// StartTunnel spawns the given tunnel type for the specified port.
func (m *TunnelManager) StartTunnel(typ TunnelType, port int) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	if _, ok := m.tunnels[typ]; ok {
		return nil // already running
	}

	var cmd *exec.Cmd
	switch typ {
	case TunnelNgrok:
		cmd = exec.Command("ngrok", "http", fmt.Sprint(port), "--log=stdout")
	case TunnelTailscale:
		cmd = exec.Command("tailscale", "funnel", "--bg", fmt.Sprint(port))
	case TunnelFunnel:
		cmd = exec.Command("cloudflared", "tunnel", "--no-autoupdate", "--url", fmt.Sprintf("http://localhost:%d", port))
	default:
		return fmt.Errorf("unknown tunnel type: %s", typ)
	}

	stdout, _ := cmd.StdoutPipe()
	stderr, _ := cmd.StderrPipe()

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("start %s: %w", typ, err)
	}

	entry := &activeTunnel{cmd: cmd}
	m.tunnels[typ] = entry

	// Parse URL from output
	scanner := func(pipe *os.File) {
		s := bufio.NewScanner(pipe)
		for s.Scan() {
			line := s.Text()
			if pat, ok := urlPatterns[typ]; ok {
				if m := pat.FindString(line); m != "" && m != entry.url {
					entry.url = m
					fmt.Fprintf(os.Stderr, "[tube] Tunnel %s: %s\n", typ, m)
					m.emit()
				}
			}
		}
	}

	if stdout != nil {
		go scanner(stdout.(*os.File))
	}
	if stderr != nil {
		go scanner(stderr.(*os.File))
	}

	go func() {
		cmd.Wait()
		m.mu.Lock()
		delete(m.tunnels, typ)
		m.mu.Unlock()
		m.emit()
	}()

	fmt.Fprintf(os.Stderr, "[tube] Tunnel %s: started\n", typ)
	return nil
}

// StopTunnel stops the given tunnel type.
func (m *TunnelManager) StopTunnel(typ TunnelType) {
	m.mu.Lock()
	defer m.mu.Unlock()

	if entry, ok := m.tunnels[typ]; ok {
		entry.cmd.Process.Signal(os.Interrupt)
		go func() {
			entry.cmd.Wait()
		}()
		delete(m.tunnels, typ)
		fmt.Fprintf(os.Stderr, "[tube] Tunnel %s: stopped\n", typ)
	}
	m.emit()
}

// Status returns the current tunnel state.
func (m *TunnelManager) Status() TunnelStatus {
	m.mu.Lock()
	defer m.mu.Unlock()

	s := TunnelStatus{URLs: make(map[string]string)}
	for typ, entry := range m.tunnels {
		switch typ {
		case TunnelNgrok:
			s.Ngrok = true
		case TunnelTailscale:
			s.Tailscale = true
		case TunnelFunnel:
			s.Funnel = true
		}
		if entry.url != "" {
			s.URLs[string(typ)] = entry.url
		}
	}
	return s
}

func (m *TunnelManager) emit() {
	if m.onChange != nil {
		m.onChange(m.Status())
	}
}
